### PR TITLE
TIP-670: Metric as immutable object value

### PR DESCRIPTION
**Pull request by damien-carcel**
**Description**

The metric object is to be independent from the storage, and only serve as an intermediate value object, not linked to the Doctrine layer anymore.

So the metric becomes an immutable object (no setters), created thanks to the MetricFactory everywhere.

Handling of the new metric ProductValue will be done later, in a separate PR.

**Definition Of Done**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | :white_check_mark:
| Added Behats                      | :negative_squared_cross_mark:
| Changelog updated                 | :negative_squared_cross_mark:
| Review and 2 GTM                  | :white_check_mark:
| Micro Demo to the PO (Story only) | :negative_squared_cross_mark:
| Migration script                  | :negative_squared_cross_mark:
| Tech Doc                          | :negative_squared_cross_mark:

### CODE REVIEW

File: src/Pim/Bundle/VersioningBundle/spec/UpdateGuesser/ProductValueUpdateGuesserSpec.php [hunk 1]
```diff
@@ -45,19 +45,19 @@ function it_marks_product_as_updated_when_a_product_price_is_removed(
...
-        ProductValueInterface $value,
-        MetricInterface $metric
-    ) {
-        $metric->getValue()->willReturn($value);
-        $value->getEntity()->willReturn($product);
-        $em->getUnitOfWork()->willReturn($unitOfWork);
-        $unitOfWork->getEntityState($product)->willReturn(UnitOfWork::STATE_MANAGED);
-        $this->guessUpdates($em, $metric, UpdateGuesserInterface::ACTION_DELETE)->shouldReturn([$product]);
-    }
+//    function it_marks_product_as_updated_when_a_product_metric_is_removed(
```
Comment (jjanvier) [inline src/Pim/Bundle/VersioningBundle/spec/UpdateGuesser/ProductValueUpdateGuesserSpec.php]: delete it if you don't want it anymore :)

File: src/Pim/Bundle/VersioningBundle/spec/UpdateGuesser/ProductValueUpdateGuesserSpec.php [hunk 2]
```diff
@@ -84,19 +84,19 @@ function it_marks_product_as_updated_when_a_product_price_is_updated(
...
-    ) {
-        $metric->getValue()->willReturn($value);
-        $value->getEntity()->willReturn($product);
-
-        $em->getUnitOfWork()->willReturn($uow);
-        $uow->getEntityChangeSet($metric)->willReturn(['data' => ['20', '25']]);
-
-        $this->guessUpdates($em, $metric, UpdateGuesserInterface::ACTION_UPDATE_ENTITY)->shouldReturn([$product]);
-    }
+//    function it_marks_product_as_updated_when_a_product_metric_is_updated(
```
Comment (jjanvier) [inline src/Pim/Bundle/VersioningBundle/spec/UpdateGuesser/ProductValueUpdateGuesserSpec.php]: same :)

File: src/Pim/Component/Catalog/Factory/MetricFactory.php [hunk 3]
```diff
@@ -2,38 +2,104 @@
...
+            $convertedData = $this->measureConverter->setFamily($family)->convertBaseToStandard($unit, $data);
+        } catch (MeasureException $e) {
+            throw new \LogicException(
+                sprintf(
+                    'Metric data %d cannot be converted to base data for unit %s and family %s',
+                    $data,
+                    $unit,
+                    $family
+                ),
+                0,
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Factory/MetricFactory.php]: $e->getCode()

;)

File: src/Pim/Component/Catalog/Model/AbstractMetric.php [hunk 4]
```diff
@@ -153,29 +85,9 @@ public function getFamily()
...
-        $this->value = $value;
-
-        return $this;
-    }
-
-    /**
-     * {@inheritdoc}
-     */
-    public function getValue()
+    public function isEqual(MetricInterface $metric)
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/AbstractMetric.php]: :+1: I like that :)

Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/AbstractMetric.php]: don't know if we should stick to Java with `equals`
or use `isEquals`
let see what other reviewers will say

File: src/Pim/Component/Catalog/Updater/Setter/MetricAttributeSetter.php [hunk 5]
```diff
@@ -63,7 +63,7 @@ public function setAttributeData(
         $this->checkData($attribute, $data);
 
         $unit = $data['unit'];
-        $amount = $data['amount'];
+        $amount = null !== $data['amount'] ? (float) $data['amount'] : null;
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/Setter/MetricAttributeSetter.php]: currently double are fetched and saved as strings
which means, if you have a really big double (as a string), maybe you'll loose precision by casting it to float (see http://php.net/manual/en/language.types.float.php)

so, not sure we should cast to float
for the moment I'd remove it

we'll tackle this problem later and globally
I've seen other (float) cast, please remove them too

File: src/Pim/Component/Catalog/spec/Updater/Copier/MetricAttributeCopierSpec.php [hunk 6]
```diff
@@ -107,9 +104,13 @@ function it_copies_a_metric_value_to_a_product_value(
         $product4->getValue('fromAttributeCode', $fromLocale, $fromScope)->willReturn($fromProductValue);
         $product4->getValue('toAttributeCode', $toLocale, $toScope)->willReturn($toProductValue2);
 
-        $metricFactory->createMetric('Weight')->shouldBeCalledTimes(1)->willReturn($metric);
+        // TODO: Why 3 times for 4 products?
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/spec/Updater/Copier/MetricAttributeCopierSpec.php]: TODO :)

File: src/Pim/Component/Catalog/Model/AbstractMetric.php (hunk 4 above)
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Model/AbstractMetric.php]: It does not really matter as long as it's consistent everywhere 🎉

Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/AbstractMetric.php]: Either `equals` or `isEqual`. `isEquals` is not grammatically correct ;).

File: src/Pim/Bundle/CatalogBundle/Resources/config/model/doctrine/Metric.mongodb.yml [hunk 7]
```diff
@@ -1,15 +0,0 @@
-Pim\Component\Catalog\Model\Metric:
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Resources/config/model/doctrine/Metric.mongodb.yml]: How to handle that regarding BC? Describe it in the Changelog, or only migration script?

Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/model/doctrine/Metric.mongodb.yml]: don't describe this one right now
it's part of a much bigger topic :p

Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Resources/config/model/doctrine/Metric.mongodb.yml]: Perfect, that's what I thought.

File: src/Pim/Component/Catalog/Model/AbstractMetric.php [hunk 8]
```diff
@@ -49,25 +46,20 @@
...
+     * @param string $baseUnit
+     * @param string $baseData
      */
-    public function setId($id)
+    public function __construct($family, $unit, $data, $baseUnit, $baseData)
     {
-        $this->id = $id;
-
-        return $this;
+        $this->family = $family;
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/AbstractMetric.php]: measureFamily

File: src/Pim/Component/Catalog/Factory/MetricFactory.php [hunk 9]
```diff
@@ -2,38 +2,104 @@
...
      *
-     * @return \Pim\Component\Catalog\Model\MetricInterface
+     * @return MetricInterface
      */
-    public function createMetric($family)
+    public function createMetric($family, $unit, $data)
     {
-        $metric = new $this->metricClass();
-        $metric->setFamily($family);
+        if (null !== $data && !is_numeric($data)) {
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Factory/MetricFactory.php]: To remove. This should be done somewhere else, with validation.
To discuss where exactly.

Comment (damien-carcel) [inline src/Pim/Component/Catalog/Factory/MetricFactory.php]: TIP card about it

Comment (damien-carcel) [inline src/Pim/Component/Catalog/Factory/MetricFactory.php]: Maybe the conversion is not to be done here… We can do that only if validation is done before.

File: src/Pim/Component/Catalog/Factory/MetricFactory.php [hunk 10]
```diff
@@ -2,38 +2,104 @@
...
+     *
+     * @throws \LogicException
+     * @return double
+     */
+    protected function convertDataToBaseData($family, $unit, $data)
+    {
+        try {
+            $convertedData = $this->measureConverter->setFamily($family)->convertBaseToStandard($unit, $data);
+        } catch (MeasureException $e) {
+            throw new \LogicException(
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Factory/MetricFactory.php]: add a todo here, replace the \logic by a custom exception

File: src/Pim/Bundle/CatalogBundle/Resources/config/converters.yml [hunk 11]
```diff
@@ -6,3 +6,4 @@ services:
         class: '%pim_catalog.converter.metric.class%'
         arguments:
             - '@akeneo_measure.measure_converter'
+            - '@pim_catalog.factory.metric'
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Resources/config/converters.yml]: Add rhis to BC

File: src/Pim/Component/Catalog/Converter/MetricConverter.php [hunk 12]
```diff
@@ -45,12 +49,16 @@ public function convert(ProductInterface $product, ChannelInterface $channel)
                 if (null === $data->getData()) {
                     return;
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Converter/MetricConverter.php]: continue
Problem is on master too (and probably previous versions).

File: src/Pim/Component/Catalog/Model/MetricInterface.php [hunk 13]
```diff
@@ -12,43 +12,51 @@
...
-     * Get family
+     * Gets the measure family of the metric.
      *
      * @return string
      */
     public function getFamily();
 
     /**
-     * To string
+     * Compares a metric (amount and unit) with the current one.
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/MetricInterface.php]: Check if the metric is equal to another one.

File: src/Pim/Component/Catalog/Model/AbstractMetric.php (hunk 8 above)
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/AbstractMetric.php]: Aborted, as it implies to also change the name of the getter and its usage everywhere...

File: src/Pim/Bundle/CatalogBundle/Resources/config/converters.yml (hunk 11 above)
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Resources/config/converters.yml]: Done.

**Comment by damien-carcel** [comment]
Approved during mob review by @phaseinducer @ahocquard @juliensnz.

Will merge once integration tests are fixed on master and added to travis.

**Pull request merged**